### ISSUE TITLE
RFE: additional Fl_Rect features

### ISSUE THREAD

**Issue by erco77** [fltk/fltk#644]
Some additional `Fl_Rect` ideas:

- `fl_push_clip()` should be able to take `const Fl_Rect&` as an argument
- `bool is_inside(x,y)` to test if (x,y) is inside the rect
- `bool intersects()`  to test if the argument intersects the rect. Argument could be a widget, xywh, or another Fl_Rect

There's probably many more, but these jump out as useful things for what I'm working on.

**Comment by Albrecht-S** [comment]
When I added `Fl_Rect` my intention was to simplify code and reduce the number of parameters, for instance for arrow drawing and similar functions. The initial implementation was minimal but the expectation was that we'd extend it.

Therefore my view is that everything that simplifies (other) code in the library can and should be added to `Fl_Rect`, particularly stuff that deals with rectangles anyway.

Some thoughts:

1. +1 on `void fl_push_clip(const Fl_Rect&)`

2. I'm not sure about `bool is_inside(x,y)`. FLTK tradition is that this should return an `int`, but I believe we allow meanwhile `bool` as well, so I'm `+/- 0` **_on the return type_** `bool`, but see below...
 
3. I looked for a precedence but I could find `is_inside` only in menu related code (where it is local and used in the same meaning). However `Fl_Rect::is_inside(...)` makes me believe (when reading it, intuitively) it should return true if `this` rectangle is _inside something else_, maybe another rectangle. I think "contains" would be a better name, and FLTK 2 had this method:
```
  /*! True if rectangle contains the pixel who's upper-left corner is at x,y */
  bool contains(int x, int y) const {return x>=x_ && y>=y_ && x<x_+w_ && y<y_+h_;}
```
The documentation is a little strange ("upper left corner" of a pixel?) but we could use it basically.

As a precedence for `contains`we have:
`  int Fl_Widget::contains(const Fl_Widget *o) const`

which has a different meaning (using the widget hierarchy) but I believe we can use the same method name for `Fl_Rect` anyway, even with more variants like
```
  int Fl_Rect::contains(int x, int y); //< this rectangle contains point (x, y) [2]
  int Fl_Rect::contains(int x, int y, int w, int h); //< this rectangle contains rectangle (x, y, w, h) [1]
  int Fl_Rect::contains(const Fl_Rect &R); //< this rectangle contains rectangle R [1]
  int Fl_Rect::contains(const Fl_Widget &W); //< this rectangle contains widget W [1]
```

[1] the return value (int) would return whether it is totally outside (0), intersects (1), or is totally inside (2) `this` Fl_Rect, similar to [fl_not_clipped()](https://www.fltk.org/doc-1.4/group__fl__drawings.html#ga2bd899d466196ffc26cc61542ee10539).

[2] would return either 0 (not contained) or 2 (totally inside), equivalent to (x, y, 1, 1)

4. `bool intersects()` would be covered by `int contains()`.

These are only my thoughts, I'm of course open for discussion and other proposals.

**Comment by Albrecht-S** [comment]
Correction: the return values 1 and 2 of `fl_not_clipped()` would be the other way around than previouosly written above (on X11)
- 1 if it is entirely inside the rectangle
- 2 if it intersects partially

These values would also be more "natural" for a point which would either return 0 or 1 (never 2).

**Comment by MatthiasWM** [comment]
Hmm, I would like to get some more global 1.5 decision in this matter. Just grep'ing for "int x, int y, int w, int h" gives my 74 hits in the header files. So we have a lot of function calls that could also use (and benefit from?) the Fl_Rect interface. If we offer this option, we should probably offer it as an alternative for *all* these functions, right?

And there are functions that return rectangles, like fl_clip_box. Fl_Widget could also return Fl_Rect on top of the rather annoying x(), y(), w(), h(). We could remove all the `draw_x = (w->as_window()) ? 0 : w->x())` code by returning different rectangles depending on the use case (window rect, drawing rect, inset drawing rect).

So, yes, I think that using Fl_Rect more often is a good idea, and adding basic logic to Fl_Rect would be really useful and reduce some amount of redundant code, but I would also then replace the driver interface with Fl_Rect everywhere. We could also gosh far and deprecate all xywh APIs (but keep it around for decades for back compatibility).

So, um, go all the way? Or leave things the way they are? Or anything in between?

**Comment by erco77** [comment]
@MatthiasWM Sounds nice to replace the Fl_Rect stuff, and would be nice if it improves the public API (in terms of simplicity). 

Also, I imagine it would centralize/simplify all the redundant code techniques for detecting rectangle intersection/clipping/etc.

Curious if it makes sense to offer Fl_Rect as an overload for Fl_Widget constructors, so that Fl_Widget(int x,int y, int w, int h ..)  can also be constructed with Fl_Widget(Fl_Rect r), and that would simply revert to calling Fl_Widget(x,y,w,h) to ensure old code that derives from widgets don't suffer from needing similar overloads, unless they want to add it. Not sure what kind of trouble that might cause, but if none, might be nice.

**Comment by erco77** [comment]
@Albrecht-S [Deleted my previous comments about is_inside() and contains() vs overlapping etc, as some of that was not taking into account the context clearly. Been a while since I thought about this stuff.]

Regarding contains(), this might just be me, but offhand I think in a subtle way contains() implies hierarchy (a 'container widget'), whereas is_inside() would not have that implication, and might be more a 2D term, like "intersect" and "overlapping". Again, might just be me.

I'm just thinking of one developer reading another dev's code and seeing e.g. `'r1->contains(r2);`' and thinking if r1/r2 are widgets (and not Fl_Rect's), that _/might/_ be a widget hierarchy test instead of a 2d operation.

**Comment by RokerHRO** [comment]
maybe you can choose different method names with clear / well documented semantics: 
* `bool Fl_Rect::contains(const Fl_Rect& other)` — returns `true` if other is completely within this rectangle. Maybe `encloses()` is a better name?
* bool Fl_Rect::is_enclosed_by(const Fl_Rect& other))` — same but with swapped roles

which semantics makes more sense to a native speaker?

vs.

*  `bool Fl_Rect::intersects(const Fl_Rect& other)` — returns `true` if both rectangles have at least 1 pixel in common 
*  `Fl_Rect Fl_Rect::intersection(const Fl_Rect& other);` — returns the intersection rectangle or a "null rectangle" if intersects(other) returns `false`?

just rough ideas…

**Comment by MatthiasWM** [comment]
@erco77 

True, I would make `Fl_Widget(const Fl_Rect &rect, const char *label=nullptr)` the default an inline `Fl_Widget(int x, int y, int w, int h, const char *label=nullptr) : Fl_Widget(Fl_Rect(x, y, w, h), label) { }`.

Thanks to C++11 constructors can now call each other. Also, we just pass a single pointer instead of four integers.

@RokerHRO 

I would like that. I we can somehow keep back compatibility.

**Comment by RokerHRO** [comment]
@MatthiasWM : Just a matter of style: I'd strongly prefer the "C++ style" and put the `&` and `*` in declarations to the _type_, not to the variable name:  `Foo* bar(Foo* f, Bar& b)` and **not** `Foo *bar(Foo *f, Bar &b);` 

I know, FLTK is not consistent here and has a lot of  "legacy" code, e.g. `FL/Fl_Browser.H` is a wild and ugly mixture of both styles, but in new code I'd favor C++ style as shown in the _ISO C++ Core Guidelines_.

**Comment by MatthiasWM** [comment]
@RokerHRO this discussion is as old as C++. The `*` modifies the use of the type IMHO, so it belongs with the variable. For example, in the line `Fl_Widget* a, b, c;`, `a` has a different type than `b` and `c`, even though they are written the same way. I know, modern code simply doesn't declare multiple variables in the same line ;-) .

But in the end, I'll be happy to change my code to what the [CMP](https://www.fltk.org/cmp.php) says.

LOL: "wild and ugly mixture", very true. Some code is thirty plus years old. We discussed a complete reformat many times over the years, but we would lose all git history, and decided against it back then. Doesn't mean it can change at some point.

**Comment by erco77** [comment]
@MatthiasWM Yes, I remember Doug Roble pointing that `Fl_Widget* a,b,c;` case, and suggesting against it as a DD coding style preference, instead keeping '*' next to each variable, to avoid that confusion. This was on one of those internal digital domain software department mailing lists, IIRC.

The exceptions for putting the '\*' with the type IMHO is when there is no variable, e.g. declaring the return type for a function, e.g. `Fl_Widget* foo_bar(..)` where it seems to make sense the '*' is next to the type.

**Comment by RokerHRO** [comment]
* https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#rl-ptr 
in combination with
* https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#res-name-one

CMP does not mention this issue, yet.

**Comment by MatthiasWM** [comment]
Yes, thanks for the links. I have read similar discussions. It's a really wide spread between a 90s code base and modern coding style. I had a hard time reading new code, and it's even harder to abandon old habits.

What is your opinion on the new(er) "auto top_window() -> Fl_Window*;" syntax? It seems to be the standard to be taught in schools now.

**Comment by RokerHRO** [comment]
> Yes, thanks for the links. I have read similar discussions. It's a really wide spread between a 90s code base and modern coding style. I had a hard time reading new code, and it's even harder to abandon old habits.

I see.

> What is your opinion on the new(er) "auto top_window() -> Fl_Window*;" syntax? It seems to be the standard to be taught in schools now.

I don't like it, it differs too much from "traditional" style for my taste. ;-)

I see benefits in function templates when the return type depends on the arguments. And small inline functions I might define them with no explicitly given return type at all, such as

```
auto get_window() { return _window; }
auto begin() { return _container.begin(); }
```

or the like, but the "trailing return type" syntax you mentioned… I don't prefer it and I'd only use it when necessary. Moreover, when things get more complicated it is even harder to _see_ the return type at first glance:

```
    auto debug(ostream& os=cout) const ->ostream& override final;
```

**Comment by erco77** [comment]
Another Fl_Rect thing that could be cool:
```
Fl_Rect& Fl_Widget::xywh() { .. }    // Fl_Widget method to return xywh as an Fl_Rect
```
This would allow simplification of a common coding pattern of setting a widget's clip region within its box(), e.g.
```
void SomeWidget::draw() {
  // Get area inside SomeWidget's box(), use as clip region for drawing
  int cx = x() + Fl::box_dx(box());
  int cy = y() + Fl::box_dy(box());
  int cw = w() - Fl::box_dw(box());
  int ch = h() - Fl::box_dh(box());
  fl_push_clip(cx,cy,cw,ch);
  ...
```
..to be instead:
```
void SomeWidget::draw() {
    // Get area inside SomeWidget's box(), use as clip region for drawing
    Fl_Rect clip(xywh());    // Fl_Widget's xywh
    clip.inset(box());       // inset box()
    fl_push_clip(clip);
    ...
 ```
Also, if this method were added as an Fl_Rect ctor:
```
Fl_Rect::Fl_Rect(const Fl_Rect& rect, Fl_Boxtype bt) { .. }  // ctor with rect + inset() box()
```
..then the pattern could I think be reduced to:
```
void SomeWidget::draw() {
    Fl_Rect clip(Fl_Rect(xywh(), box()));
    fl_push_clip(clip);
    ...
```
..or I suppose even:
```
void SomeWidget::draw() {
    fl_push_clip(Fl_Rect(xywh(), box()));   // one liner!
    ...
```
_(This came to me at 4am, not sure if it's a good idea or not. I'm goin back to bed)_

**Comment by gvnnz** [comment]
> Another Fl_Rect thing that could be cool:
> 
> ```
> Fl_Rect& Fl_Widget::xywh() { .. }    // Fl_Widget method to return xywh as an Fl_Rect
> ```

Great! Small nitpick: I'd go with `as_rect()` instead of `xywh()` to make it clearer we are getting an `Fl_Rect` back.

**Comment by RokerHRO** [comment]
> > Another Fl_Rect thing that could be cool:
> > ```
> > Fl_Rect& Fl_Widget::xywh() { .. }    // Fl_Widget method to return xywh as an Fl_Rect
> > ```
> 
> Great! Small nitpick: I'd go with `as_rect()` instead of `xywh()` to make it clearer we are getting an `Fl_Rect` back.

…or use language feature instead of introducing another naming convention:
```
explicit Fl_Widget::operator Fl_Rect& const();
```

or even more radical: Just derive `Fl_Widget` from `Fl_Rect`, so every Widget can be used automatically where a rectangle is expected.

**Comment by erco77** [comment]
Lots of good comments.. curious for devs to weigh in.

Offhand a few replies:

- `as_rect()` sounds better than my `xywh()` sugg - I like it! Would also read clearer in code, e.g. `Fl_Rect clip = w->as_rect().inset(box());`. I figure most toolkits would use `as_xyz()` naming. 

- The suggestion of simply deriving `Fl_Widget` from `Fl_Rect` is very seductive. I imagine had FLTK been designed with `Fl_Rect` in mind, that might have been "the right way"! As a retrofit tho, might be some issues. Surely it breaks ABI, but not a big deal because we can simply delay implementation to a major release. But mainly,not sure `Fl_Rect` was designed for this kind of inheritance; some methods like `inset()` might be dangerous to use within widget code, as it would permanently adjust Fl_Widget's internal xywh values. It seems Fl_Rect works better outside of Fl_Widget, or perhaps as an internal private variable.

**Comment by RokerHRO** [comment]
@erco77 : Yes, making `Fl_Rect` such a "super base class" might have unforeseeable – and perhaps dangerous? – side effects. But what about my suggestion with the (explicit) conversion operator?

**Comment by MatthiasWM** [comment]
For every widget, there are three rect areas that are interesting:

1: most often used: the drawing area. For regular widgets, that is x, y, w, h, but for windows and sub windows, it is 0, 0, w, h. I propose `Fl_Rect Fl_Widget::drawing_rect();`

2: many widgets use inset, the drawing area after drawing the box type. I propose `Fl_Rect Fl_Widget::inset_rect();` or `boxed_rect`.

3: the actual widget rectangle, x, y, w, h, is rarely used (see window/sub window limitation), and I propose `Fl_Rect Fl_Widget::inset_rect();` to return that one, or `const Fl_Rect& Fl_Widget::rect();` if we decide to store the widget coordinates as a rect anyway and inline x(), y(), w(), and h().

All the other drawing functions that use rectangular coordinates can then follow as needed.

**Comment by erco77** [comment]
@MatthiasWM Would your `inset_rect(void)` proposal imply `box()` as the inset value, in which case maybe including the word 'box' in the method name would help? e.g. `Fl_Rect inset_box()`. Would be good if it supported `int` and `Fl_Boxtype` values to be explicit, as there might be reasons to access the Fl_Rect well within the widget bounds, such as to leave room for "margins" and such.

@RokerHRO Regarding the `explicit / operator` suggestion, TBH I wasn't sure what that does or how it's used, so I can't weigh in. (I'm still in the 90s lol)

**Comment by RokerHRO** [comment]
@erco77 : Being an "old-timer" is nothing bad. Unwilling to learning new would be bad. So as long as you are wiling (and able) to learn, everything is fine: https://en.cppreference.com/w/cpp/language/cast_operator.html

**Comment by MatthiasWM** [comment]
@erco77 it would use the current box type. The idea is to replace this: void 
```
Fl_Chart::draw() {
  draw_box();
  Fl_Boxtype b = box();
  int xx = x() + Fl::box_dx(b);
  int yy = y() + Fl::box_dy(b);
  int ww = w() - Fl::box_dw(b);
  int hh = h() - Fl::box_dh(b);
  fl_push_clip(xx, yy, ww, hh);
```
with 
```
void Fl_Chart::draw() {
  draw_box();
  auto r = inset_rect();
  fl_push_clip(r);
```
I found spontaneously about 40 instances where this would apply.

I like `rect` in the name to indicate what it returns. `boxed_rect()`?

An if you need the inset for a different box type, we can have `Fl_Rect Fl_Widget::inset_rect(Fl_Boxtye bt=-1);` or something.

@RokerHRO in this specific case I would not cast to Fl_Rect because it is not clear which of the three rectangles is returned.

**Comment by RokerHRO** [comment]
@MatthiasWM : Okay, in that case a method with meaningful name is a lot better than a cast operator. :-)

**Comment by Albrecht-S** [comment]
> I like rect in the name to indicate what it returns. boxed_rect()?

FTR, concerning method names, we have already:
1. [void Fl_Scroll::bbox(int &, int &, int &, int &) const](https://www.fltk.org/doc-1.4/classFl__Scroll.html#ab0448d35fdc00bbabaab1c6b9b2af525)

> Returns the bounding box for the interior of the scrolling area, inside the scrollbars.

which is IMHO not a good name but a useful method.

2. [void Fl_Tabs::client_area(int &rx, int &ry, int &rw, int &rh, int tabh=0)](https://www.fltk.org/doc-1.4/classFl__Tabs.html#a340c3c801614767966a60992fff76a14)

I suggest to define

1. `Fl_Rect Fl_Widget::box_rect() const` to return the entire box area based on x(), y(), w(), h().
2. `Fl_Rect Fl_Widget::client_rect() const` to return the entire area inside the box (frame).

Regarding (1): I can also imagine to call it just `Fl_Rect Fl_Widget::rect() const` which is even simpler and describes exactly what it does (returns).

Regarding (2): `Fl_Widget::client_rect()` would admittedly be semantically different than `Fl_Tabs::client_area()` because it would be static (depends only on the box type) rather than `Fl_Tabs::client_area()` which is dynamic (depends on current children sizes). An alternate name could be inset_rect() as proposed by Matthias.

An alternate name for (2) could be `Fl_Rect Fl_Widget::clip_rect() const` with the same semantics as `client_rect()`. If this was `virtual` it could be overridden by subclasses, but this could become confusing: I'd rather like to have a method that returns the same for all classes.

Another approach would be to define `virtual Fl_Rect Fl_Widget::rect(int what = 0) const` where `what` would be a number that determines what exactly to return, for instance:
- 0: entire box (default)
- 1: inside area of box (clipping box)
- 2: area of children inside the widget, similar to `Fl_Tabs::client_area(...)`. Subclasses could override this for their own purposes.
- \>9 (or even higher): free to use in subclasses

Maybe there are more use cases (I didn't investigate code), but the version above looks to be very extensible.

**Comment by MatthiasWM** [comment]
Here is my suggestion to get this process started. This covers 95% of use cases. I avoided the word "region" because in GUI programming, a Region is already fixed term.

As for @erco77 s example, the code to clip the drawing are could then be:
```
void SomeWidget::draw() {
    fl_push_clip(local_rect());
    ...
```

Any comments?

```
/**
  Returns the widget's outer rectangle in its parent’s coordinate system.

  For Fl_Window and derived classes, the position is given relative to
  the parent window or the desktop.

  The returned rectangle corresponds to x(), y(), w(), and h().

  \return the widget’s bounding rectangle as an Fl_Rect
  \see local_rect() for the widget’s full drawable area in local coordinates
  \see client_rect() for the inner (safe) drawing area
*/
Fl_Rect Fl_Widget::rect() const;

/**
  Returns the widget’s full drawable area in local coordinates.

  The returned rectangle always has its origin at (0, 0) and the same
  width and height as the widget, without accounting for frame
  decorations such as box() borders or scroll bars.

  \return an Fl_Rect representing the undecorated local drawing area
  \see rect() for the widget’s outer rectangle in parent coordinates
  \see client_rect() for the inner (safe) drawing area
*/
Fl_Rect Fl_Widget::local_rect() const;

/**
  Returns the widget’s inner (safe) drawing area in local coordinates.

  This rectangle excludes frame decorations defined by box() and omits
  any scroll bars or similar UI elements provided by subclassed widgets.
  It represents the area where content can be safely drawn.

  \return an Fl_Rect representing the safe client area
  \see rect() for the widget’s outer rectangle
  \see local_rect() for the full drawable area before insets
*/
Fl_Rect Fl_Widget::client_rect() const;
```

**Comment by Albrecht-S** [comment]
> Any comments?

👍 but with the following comments. I'm not sure what the first two methods are good for, or better: what distinguishes these two methods.

1. Fl_Rect Fl_Widget::rect() const;
> For Fl_Window and derived classes, the position is given relative to the parent window or the desktop.

I don't see a use case for "or the desktop", particularly because we don't get top level window coordinates relative to the desktop on Wayland. Hence this would fail by design on Wayland. IMHO we should always use coordinates relative to the parent window of the widget - unless the widget is a top-level window. For the latter we should set `x` and `y` to (0, 0).

2. Fl_Rect Fl_Widget::local_rect() const;
> The returned rectangle always has its origin at (0, 0) and the same width and height as the widget, without accounting for frame decorations such as box() borders or scroll bars.

What does "_The returned rectangle always has its origin at (0, 0)_" mean? What is the origin of a rectangle? Does this mean that it is relative to its parent window, but the (x,y) coordinates are zero for top-level windows? It the latter is true, then this would be the same as 1. if my comment above was taken into account.

3. Fl_Rect Fl_Widget::client_rect() const;
> This rectangle ... and omits any scroll bars or similar UI elements provided **by subclassed widgets**.

The latter implies that this method has to be virtual and must be overridden by subclasses. Is this what you meant?

That said, I'm not sure about use cases for the distinction of 1. and 2., or if we (should) end up having only one of them with clear (and clean) semantics and documentation. Or did I miss anything?

**Comment by MatthiasWM** [comment]
1: it's back compatible to what using x(), y(), w(), h() does. It will probably be rarely used, but it's helpful for initial porting. I will rethink that. 2: Oops, the text is wrong. (0, 0) is only true for the local rect of Fl_Windows. I must have mixed up the docs of Fl_Widget::local_rect and Fl_Window::local _rect. 3: it's very useful because it delivers the area the usable area without interfering with the frame. Yes, it must be virtual because we do special case stuff in Tabs, Scroll, and some other widgets.

Yes, we can probably remove case 1 and expect users to simply call x() and y() instead for Windows, even if that has less meaning on Wayland. But even there it can still be used to position a toolbox or popup menu relative to a window.

**Comment by Albrecht-S** [comment]
@MatthiasWM Thanks for your reply.
> 1: ... I will rethink that.
> I must have mixed up the docs of Fl_Widget::local_rect and Fl_Window::local _rect. ...

??? It would be great if you could sort that out and post a better API proposal so we can discuss this finally. I think that we can use one of 1. and 2. and of course we can use 3: `virtual Fl_Rect Fl_Widget::client_rect() const;`. I only wanted to clarify that this one needs to be virtual (as you confirmed).

Besides that, the goal should be to make the example code
```C++
void SomeWidget::draw() {
    fl_push_clip(local_rect());
    ...
```
work, whatever the name of the method may be. For this to work we definitely need a method that uses window-local coordinates with (0, 0) if the widget is a top-level window itself or any widget that is positioned in the top left corner of a window.

**Comment by MatthiasWM** [comment]
Updated proposal, based on @Albrecht-S s suggestions.

It is my goal to make drawing code as simple and as readable as possible:
```
void SomeWidget::draw() {
    fl_push_clip(client_rect());
    ...
```

```
/**
  Removed Fl_Rect Fl_Widget::rect() const;
  The widget rect (and x(), and y()) has a different meaning for widgets
  (the drawing area) and for windows (the position on the desktop, if
  supported) and sub windows (the position within the parent)
  and is relatively useless as a shortcut. Use x(), y(), and w(), h() instead.
 */

/**
  Returns the widget’s full drawable area in local coordinates.
  (note: name upgraded from "local_rect" to "rect" as it seems to become
  a very commonly used function)
  \return an Fl_Rect representing the undecorated drawing area
  \see client_rect() for the inner (safe) drawing area
*/
Fl_Rect Fl_Widget::rect() const;

/**
  Returns the widget’s inner (safe) drawing area in local coordinates.

  This rectangle excludes frame decorations defined by box() and omits
  any scroll bars or similar UI elements. It represents the area where 
  content can be safely drawn.

  \return an Fl_Rect representing the safe client area
  \see rect() for the widget’s outer rectangle
  \see local_rect() for the full drawable area before insets
*/
virtual Fl_Rect Fl_Widget::client_rect() const;
```

**Comment by erco77** [comment]
Sounds good, but I wouldn't have guessed "client" meant "inner".

If the widget rect is called `rect()`, would make more sense to me to call it `inner_rect()`..?

**Comment by Albrecht-S** [comment]
@erco77 wrote:
> Sounds good, but I wouldn't have guessed "client" meant "inner".

Greg, you are the native English speaker, what would you suggest to more precisely describe the nature of this "inner" rectangle? Just `inner_rect()` might be too ambiguous.

Also note that we have
```C++
void Fl_Tabs::client_area(int &rx, int &ry, int &rw, int &rh, int tabh=0);
```

where `"client"` is used in a similar but different meaning: this is the area where children of `Fl_Tabs` would be located, not the entire drawing area (rectangle).

Now, thinking about it ...
- it would be consequent and consistent to create `Fl_Rect Fl_Tabs::client_rect()` with the same semantics as `Fl_Tabs::client_area(...)` but this would conflict with the proposed `virtual Fl_Rect Fl_Widget::client_rect() const;` method.
- using another "prefix" like `inner` (`virtual Fl_Rect Fl_Widget::inner_rect() const;`) would resolve this conflict
- what about `drawing_rect()` instead? it's not ideal because the box borders are also _drawn_ but would be excluded here
- Greg, or anybody else, are there better proposals?

Finally, here's another important thought: the current proposal of client_rect documents:
> This rectangle excludes frame decorations defined by box() and omits any scroll bars or similar UI elements. It represents the area where content can be safely drawn.

Although this would be correct (if implemented properly by subclasses) it's not really helpful for such (sub)classes that manage their own scrollbars etc.. Such classes would **_also_** (first) need a method that does only "subtract" (inset) the box borders and nothing else. In contrast to the discussed method this wouldn't need to be virtual and wouldn't have to be reimplemented in subclasses.
Code example:
```C++
void SomeWidget::draw() {
    auto inner_rect = box_inner_rect(); // --> name to be discussed <--
    fl_push_clip(inner_rect);
    // manage scrollbars or something else ...
    Fl_Rect client_rect = ... ? // calculate whatever the designer of the widget needs
    // draw children ...
}
```

**My first conclusion:**  the proposed `client_rect()` method would have to be (re-)**implemented** in subclasses anyway if they have such needs. Hence it does IMHO not make much sense to declare this as a virtual method of `Fl_Widget`. Hence the needed method (to simplify clipping and more) would be just to return the rectangle inside the box borders and nothing else. No need to exclude scrollbars, let's leave this to the widget designer. KISS.

That said, the method that excludes the borders (and nothing else) wouldn't need to be virtual and could be like below. My modified full suggestion would be (most of it copied from Matthias' proposal):
```C++
/**
  Returns the widget’s full drawable area in local coordinates.
  \return an Fl_Rect representing the full drawing area including box borders
  \see client_rect() for the inner (safe) drawing area
*/
Fl_Rect Fl_Widget::rect() const;

/**
  Returns the widget’s inner (safe) drawing area in local coordinates.

  This rectangle excludes frame decorations defined by box(). 
  It represents the area where content can be safely drawn.

  \return an Fl_Rect representing the safe client drawing area
  \see rect() for the widget’s outer rectangle
*/
Fl_Rect Fl_Widget::client_rect() const;
```

Note: both are not virtual. Replace name `client_rect()` with something better like `inner_rect()` if desired.

Sorry for the long comment, the confusion, and for "turning back" now: After thinking even longer, a subclass of `Fl_Scroll` might want to know the current arrangement of scrollbars (or better: the inner space, excluding the scrollbars). Hence the subclass `My_Scroll` might want to call `Fl_Scroll::client_area()` for its own use (leaving the calculation to the parent class Fl_Scroll).

Do we need three ~classes~ methods nevertheless? Slightly confused ...

**Comment by erco77** [comment]
"inner" seems like the right term here. We're needing a geometric term, and "client" just isn't. In software contexts, "client" is pretty limited to either client/server models, or client==customer, e.g. in subscription based systems.

Bill would probably be the last word, but I'm thinking he'd surely pick "inner" over "client".

_Regarding the "three classes" case, I haven't really looked at that yet, so can't weigh in._

**Comment by erco77** [comment]
Following up re: `Fl_Tabs::client_area()`, I think client kinda works OK there because "client" is being used in a child in a parent/child relationship, and in Fl_Tabs the area is used by the children.

I'm thinking `client_area()` could have just as easily could have been called `child_area()`.

The difference with the Fl_Rect context, there's not really a parent/child relationship, it's more one of dimensional space, purely geometric.

Just in general, "client" is kinda a formal term, sort of "sales speaky"; if I'm hired by a company to do some consulting work, I'd informally describe them as a "customer", but if I wanted to impress someone, I'd say they're a "client". FLTK generally goes for informal, catchy naming.. easy to remember, quick and easily makes sense in context.

**Comment by Albrecht-S** [comment]
> Regarding the "three classes" case, I haven't really looked at that yet, so can't weigh in.

Sorry again for the confusion. I edited my comment: what I wanted to ask was about three (rather than only two) **methods** for the different rects.

**Comment by MatthiasWM** [comment]
According to ChatGPT who happened to know the entire internet by heart  (LOL):

> In GUI terminology, the area inside a push button—the part that is not the frame, border, chrome, or decoration—is typically called:
The button’s “content area.”
Other closely related terms used in various GUI frameworks:
Content area — the standard, neutral term across many toolkits.
Label area — when the content is primarily text.
Client area — used in some UI frameworks (e.g., Win32) for the drawable/usable region of a control.
Interior — sometimes used informally.
In FLTK, Qt, GTK, Cocoa, Win32, and web CSS alike, content area is the most precise and widely understood term.

So how about `content_rect()` then?

About the three methods to get a rectangle:

The first one just mirrored x(), y(), w() and h() and is relatively useless. x() and y() do different things if they are in an Fl_Window vs. an Fl_Widget. If you look at core code that needs the coordinates of an unknown widget, it's typically:
```
int X = x(), Y = y();
if (as_window()) {
  X = 0; Y = 0;
} 
fl_rectf(X, Y, w(), h());
```
Fl_Group::bounds(), Fl_Group::resize() (2x), Fl_Button::draw(), Fl_Group::draw_outside_label(), Fl_Tabs::draw(), event_inside(), and a few more in apps.

**Comment by Albrecht-S** [comment]
> So how about content_rect() then?

👍 +1 for `content_rect()`.

> About the three methods to get a rectangle: [...]

@MatthiasWM Sorry, please read my comment again, but maybe I was not clear enough. Sorry. What **I meant** was

1. The first method gets the full rectangle: `Fl_Rect Fl_Widget::rect() const;` .
2. The second method would get the full rect w/o borders: former `Fl_Rect Fl_Widget::client_rect() const;`, now maybe `Fl_Rect Fl_Widget::content_rect() const;`.
3. The third method would do what you proposed earlier: this method eclipses scrollbars etc. from the returned rectangle. I don't know how to name this, let's say `virtual Fl_Rect Fl_Widget::child_rect() const;`. This method would be **virtual** and would need to be (re-)implemented in some - but not all - widget subclasses. Suggestions for a better name appreciated. _[OT(?): What would this method return for `Fl_Tabs`?]_

All methods would return "relative" coordinates, i.e. (0,0) for (x,y) in **all** windows. This is necessary if we want to use the returned `Fl_Rect` directly for clipping and drawing.

In my previous comment I thought that we need only 1. and 2., but at the end I wanted to express that 3. might also be useful.

And yes, your [example code](https://github.com/fltk/fltk/issues/644#issuecomment-3647753285) (which we want to simplify for the caller) would now go into all these new methods to produce "relative", i.e. (0,0) based, coordinates:

```C++
/**
  Returns the widget’s full drawable area in local coordinates.

  These coordinates can directly be used for drawing, i.e. for subclasses of Fl_Window
  the \c x and \c y components are always 0 (zero).

  \return an Fl_Rect representing the full drawing area including box borders
  \see client_rect() for the inner (safe) drawing area
*/
Fl_Rect Fl_Widget::rect() const {
  if (as_window) {
    return Fl_Rect(0, 0, w(), h());
  else
    return Fl_Rect(x(), y(), w(), h());
  }
}
```

(code untested)

Looking at this implementation: I wonder if this method needs to be virtual as well because it calls `virtual as_window()`? Hmm ...
